src/vendor/azure-uamqp-c/samples/multiple_receivers_share_connection_sample/main.c

```diff
@@ -0,0 +1,255 @@
+    STRING_HANDLE sas_key_name;
+    STRING_HANDLE sas_key_value;
+    STRING_HANDLE resource_uri;
+    STRING_HANDLE encoded_resource_uri;
+    STRING_HANDLE sas_token;
+    BUFFER_HANDLE buffer;
+    TLSIO_CONFIG tls_io_config = { EH_HOST, 5671 };
+    const IO_INTERFACE_DESCRIPTION* tlsio_interface;
+    SASLCLIENTIO_CONFIG sasl_io_config;
+    time_t currentTime;
+    size_t expiry_time;
+    CBS_HANDLE cbs;
+
+    gballoc_init();
+
+    /* create the TLS IO */
+    tlsio_interface = platform_get_default_tlsio();
+    tls_io = xio_create(tlsio_interface, &tls_io_config);
+
+    /* create the SASL client IO using the TLS IO */
+    sasl_io_config.underlying_io = tls_io;
+    sasl_io_config.sasl_mechanism = sasl_mechanism_handle;
+    sasl_io = xio_create(saslclientio_get_interface_description(), &sasl_io_config);
+
+    /* create the connection, session and link */
+    connection = connection_create(sasl_io, EH_HOST, "aname", NULL, NULL);
+    connection_set_trace(connection, true);
+    session = session_create(connection, NULL, NULL);
+    session_set_incoming_window(session, 655565);
+    session_set_outgoing_window(session, 65536);
+
+    /* Construct a SAS token */
+    sas_key_name = STRING_construct(EH_KEY_NAME);
+
+    /* unfortunately SASToken wants an encoded key - this should be fixed at a later time */
+    buffer = BUFFER_create((unsigned char*)EH_KEY, strlen(EH_KEY));
+    sas_key_value = Base64_Encoder(buffer);
+    BUFFER_delete(buffer);
+    resource_uri = STRING_construct("sb://" EH_HOST "/" EH_NAME);
+    encoded_resource_uri = URL_EncodeString(STRING_c_str(resource_uri));
+
+    /* Make a token that expires in one hour */
+    currentTime = time(NULL);
+    expiry_time = (size_t)(difftime(currentTime, 0) + 3600);
+    sas_token = SASToken_Create(sas_key_value, encoded_resource_uri, sas_key_name, expiry_time);
+
+    cbs = cbs_create(session);
+    cbs_set_trace(cbs, true);
+    if (cbs_open_async(cbs, on_cbs_open_complete, cbs, on_cbs_error, cbs) == 0)
+    {
+	    (void)cbs_put_token_async(cbs, "servicebus.windows.net:sastoken", "sb://" EH_HOST "/" EH_NAME, STRING_c_str(sas_token), on_cbs_put_token_complete, cbs);
+
+	    while (!auth)
+	    {
+		    connection_dowork(connection);
+	    }
+    }
+
+    STRING_delete(sas_token);
+    STRING_delete(sas_key_name);
+    STRING_delete(sas_key_value);
+    STRING_delete(resource_uri);
+    STRING_delete(encoded_resource_uri);
+
+    int max_link_credit = 300;
+    int partition_cnt = 2;
+
+    MESSAGE_RECEIVER_HANDLE* receiver_handlers = malloc(sizeof(MESSAGE_RECEIVER_HANDLE) * partition_cnt);
+    SESSION_HANDLE* session_handlers = malloc(sizeof(SESSION_HANDLE) * partition_cnt);
+    LINK_HANDLE* link_handlers = malloc(sizeof(LINK_HANDLE) * partition_cnt);
+
+    int* running_context = malloc(sizeof(int) * partition_cnt);
+
+    for (int i = 0; i < partition_cnt; i++) {
+	    session_handlers[i] = create_session(connection);
+        LINK_HANDLE* created_link = NULL;
```

Comment:
I add more comments to help understand the code.
This c code is basically showing how to multiplex uamqp connection and use that for receiving, not targeting at showing EH functionality - I copied most part of the code from other c sample.😀

However, it can still be improved as you said, in the future time when we have more time